In [78]:
from sympy import *

# Завдання 1
## ГЕНЕРАЦІЯ СЛАР

In [79]:
n = 112
m = 1
l = 2
k = 3
i = 2
j = 1
g = 0.33

In [80]:
A = [[33.9, -0.2, 3.3, -0.8, 46.5, -0.0495],
    [31.92, 0.35, -6.3, -0.6, -4.5, 0.0495],
    [123.2, -0.5, 4.2, -11.76, 0.5, -0.2475],
    [582.4, 0.1, -8.6, 0.2, -1.25, -0.3465],
    [11.2, -45.9, 3.12, -0.8, 0.8, -0.0396],
    [0.62, 0.3, -5.4, -0.82, 2.5, 12.266]]
a = Matrix(A)
a

Matrix([
[ 33.9,  -0.2,  3.3,   -0.8,  46.5, -0.0495],
[31.92,  0.35, -6.3,   -0.6,  -4.5,  0.0495],
[123.2,  -0.5,  4.2, -11.76,   0.5, -0.2475],
[582.4,   0.1, -8.6,    0.2, -1.25, -0.3465],
[ 11.2, -45.9, 3.12,   -0.8,   0.8, -0.0396],
[ 0.62,   0.3, -5.4,  -0.82,   2.5,  12.266]])

In [81]:
b = [2, -3, 112, -1, 2, 1]
b = Matrix(b)
b

Matrix([
[  2],
[ -3],
[112],
[ -1],
[  2],
[  1]])

# Завдання 2

## ПЕРЕВІРКА УМОВИ ЗБІЖНОСТІ

In [82]:
import numpy as np
diag_values = np.diag(A)
dg = np.diag(diag_values)
visual = Matrix(dg)
visual

Matrix([
[33.9,    0,   0,   0,   0,      0],
[   0, 0.35,   0,   0,   0,      0],
[   0,    0, 4.2,   0,   0,      0],
[   0,    0,   0, 0.2,   0,      0],
[   0,    0,   0,   0, 0.8,      0],
[   0,    0,   0,   0,   0, 12.266]])

In [83]:
A_det = np.linalg.det(A)
A_det

-1036577001.0524973

In [84]:
norm_A = np.linalg.norm(A)
norm_A

601.1928249265454

In [85]:
ADG = A - dg
adg_vis = Matrix(ADG)
adg_vis


Matrix([
[    0,  -0.2,  3.3,   -0.8,  46.5, -0.0495],
[31.92,     0, -6.3,   -0.6,  -4.5,  0.0495],
[123.2,  -0.5,    0, -11.76,   0.5, -0.2475],
[582.4,   0.1, -8.6,      0, -1.25, -0.3465],
[ 11.2, -45.9, 3.12,   -0.8,     0, -0.0396],
[ 0.62,   0.3, -5.4,  -0.82,   2.5,       0]])

In [86]:
inv = -np.linalg.inv(dg).dot(ADG)
inv_vis = Matrix(inv)
inv_vis

Matrix([
[                  0, 0.00589970501474926, -0.0973451327433628,  0.023598820058997,   -1.3716814159292, 0.00146017699115044],
[              -91.2,                   0,                18.0,   1.71428571428571,   12.8571428571429,  -0.141428571428571],
[  -29.3333333333333,   0.119047619047619,                   0,                2.8, -0.119047619047619,  0.0589285714285714],
[            -2912.0,                -0.5,                43.0,                  0,               6.25,              1.7325],
[              -14.0,              57.375,                -3.9,                1.0,                  0,              0.0495],
[-0.0505462253383336, -0.0244578509701614,   0.440241317462906, 0.0668514593184412, -0.203815424751345,                   0]])

In [89]:
norm_inv = np.linalg.norm(inv, np.inf)
norm_inv

2963.4825

### Штучно перетворимо матрицю так, щоб привести А до вигляду з діагональним перевершенням.


In [90]:
abs_sum1 = round(sum(list(map(abs, A[0])))) + 2 
abs_sum2 = round(sum(list(map(abs, A[1])))) + 2 
abs_sum3 = round(sum(list(map(abs, A[2])))) + 2
abs_sum4 = round(sum(list(map(abs, A[3])))) + 2
abs_sum5 = round(sum(list(map(abs, A[4])))) + 2
abs_sum6 = round(sum(list(map(abs, A[5])))) + 2
print(abs_sum1, abs_sum2, abs_sum3, abs_sum4, abs_sum5, abs_sum6)
per_list = [abs_sum1, abs_sum2, abs_sum3, abs_sum4, abs_sum5, abs_sum6]

87 46 142 595 64 24


In [91]:
A = np.array([[33.9, -0.2, 3.3, -0.8, 46.5, -0.0495],
    [31.92, 0.35, -6.3, -0.6, -4.5, 0.0495],
    [123.2, -0.5, 4.2, -11.76, 0.5, -0.2475],
    [582.4, 0.1, -8.6, 0.2, -1.25, -0.3465],
    [11.2, -45.9, 3.12, -0.8, 0.8, -0.0396],
    [0.62, 0.3, -5.4, -0.82, 2.5, 12.266]], dtype = float)
np.fill_diagonal(A, per_list)
a_vis = Matrix(A)
a_vis


Matrix([
[ 87.0,  -0.2,   3.3,   -0.8,  46.5, -0.0495],
[31.92,  46.0,  -6.3,   -0.6,  -4.5,  0.0495],
[123.2,  -0.5, 142.0, -11.76,   0.5, -0.2475],
[582.4,   0.1,  -8.6,  595.0, -1.25, -0.3465],
[ 11.2, -45.9,  3.12,   -0.8,  64.0, -0.0396],
[ 0.62,   0.3,  -5.4,  -0.82,   2.5,    24.0]])

In [92]:
dg = np.zeros((6, 6))
np.fill_diagonal(dg, per_list)
dg_vis = Matrix(dg)
dg_vis

Matrix([
[87.0,    0,     0,     0,    0,    0],
[   0, 46.0,     0,     0,    0,    0],
[   0,    0, 142.0,     0,    0,    0],
[   0,    0,     0, 595.0,    0,    0],
[   0,    0,     0,     0, 64.0,    0],
[   0,    0,     0,     0,    0, 24.0]])

In [93]:
A_det = np.linalg.det(A)
A_det

264655423751.8004

In [94]:
norm_A = np.linalg.norm(A)
norm_A

865.2518162287555

In [95]:
ADG = A - dg
ADG_vis = Matrix(ADG)
ADG_vis

Matrix([
[    0,  -0.2,  3.3,   -0.8,  46.5, -0.0495],
[31.92,     0, -6.3,   -0.6,  -4.5,  0.0495],
[123.2,  -0.5,    0, -11.76,   0.5, -0.2475],
[582.4,   0.1, -8.6,      0, -1.25, -0.3465],
[ 11.2, -45.9, 3.12,   -0.8,     0, -0.0396],
[ 0.62,   0.3, -5.4,  -0.82,   2.5,       0]])

In [96]:
inv = -np.linalg.inv(dg).dot(ADG)
inv_vis = Matrix(inv)
inv_vis

Matrix([
[                  0,   0.00229885057471264, -0.0379310344827586, 0.00919540229885057,    -0.53448275862069, 0.000568965517241379],
[ -0.693913043478261,                     0,    0.13695652173913,  0.0130434782608696,   0.0978260869565217, -0.00107608695652174],
[ -0.867605633802817,   0.00352112676056338,                   0,  0.0828169014084507, -0.00352112676056338,  0.00174295774647887],
[ -0.978823529411765, -0.000168067226890756,   0.014453781512605,                   0,  0.00210084033613445, 0.000582352941176471],
[             -0.175,             0.7171875,            -0.04875,              0.0125,                    0,           0.00061875],
[-0.0258333333333333,               -0.0125,               0.225,  0.0341666666666667,   -0.104166666666667,                    0]])

In [97]:
norm_inv = np.linalg.norm(inv, np.inf)
norm_inv

0.9961285714285715

 # Завдання 3  
 ## Знайти розв’язки перетвореної системи  точними методами Крамера, оберненої матриці, LU – декомпозиції.

In [118]:
def cramer(A, b):
    n = A.shape[0]
    det_A = A.det()
    x = []
    for i in range(n):
        matrix = A.copy()
        matrix[:,i] = b
        det_i = matrix.det()
        x = det_i / det_A
        x.append(x.evalf())
    return x

### Щоб реалізувати метод Крамера потрібно:
### Записати систему рівнянь, де кожне рівняння представлене у вигляді лінійної комбінації змінних із коефіцієнтами.

### Обчислити визначник матриці коефіцієнтів.

### Для кожної змінної замінити відповідний стовпець матриці коефіцієнтів стовпець вільних членів і обчислити визначник отриманої матриці.

### Вирішити систему, обчисливши кожну змінну як відношення визначника, що відповідає цій змінній, до визначника матриці коефіцієнтів.


### Розв'язок методом крамера:

In [119]:
cramer_X = Matrix(cramers_rule(a_vis, b))
cramer_X

Matrix([
[-0.0369807352396914],
[ 0.0792059120634617],
[  0.825156739087997],
[ 0.0466758130317061],
[ 0.0550222528025222],
[  0.223155466999352]])

In [120]:
def inverse_method(matrix, free_terms_vector):
    inv_A = matrix.inv()
    x = (inv_A * free_terms_vector).evalf()
    return x


### Метод полягає в знаходженні оберненої матриці M^(-1) та обчисленні розв'язку x = M^(-1) b. 


### Розв'язок методом оберненої матриці

In [124]:
inverse_X = inverse_method(a_vis, b)
inverse_X

Matrix([
[-0.0369807352396914],
[ 0.0792059120634618],
[  0.825156739087997],
[ 0.0466758130317061],
[ 0.0550222528025222],
[  0.223155466999352]])

In [137]:
def LU_decomposition(A, b):
    n = A.shape[0]
    L, U, _ = A.LUdecomposition()
    y = []
    for i in range(n):
        y_i = b[i]
        for j in range(i):
            y_i -= L[i, j]*y[j]
        y.append(y_i)
    x = []
    for i in range(n-1, -1, -1):
        x_i = y[i]
        for j in range(i+1, n):
            x_i -= U[i, j]*x[n-j-1]
        x_i /= U[i, i]
        x.append(x_i)
    x.reverse()
    return x

### Розв'язок методом LU-декомпозиції

In [138]:
LU_decomposition_X = LU_decomposition(a_vis, b)
Matrix(LU_decomposition_X)

Matrix([
[-0.0369807352396914],
[ 0.0792059120634617],
[  0.825156739087997],
[ 0.0466758130317061],
[ 0.0550222528025222],
[  0.223155466999352]])

### Бачимо, що всі ров'язки співпадають.

# Завдання 4
## Скласти програми ітераційних методів Якобі, Зейделя. Провести серію обчислювальних експериментів із варіацією значень ε (ε = 0.1, ε = 0.01, ε = 0.001, ε = 10−6) і початкових наближень. Визначити залежності, які впливають на кількість ітерацій.

In [141]:
x0 = [0, 0, 0, 0, 0, 0]
epsilons = [0.1, 0.01, 0.001, 0.000001]

In [150]:
def jacobi_method(A, b, x0, epsilon):
    n = A.shape[0]
    x = x0.copy()
    iterations = 0
    while True:
        iterations += 1
        x_new = []
        for i in range(n):
            s = 0
            for j in range(n):
                if j != i:
                    s += A[i, j] * x[j]
            x_i = (b[i] - s) / A[i, i]
            x_new.append(x_i)
        if max([abs(x_new[i]-x[i]) for i in range(n)]) < epsilon:
            break
        x = x_new
    return x, iterations


### Розв'язки методом Якобі:

In [148]:
for epsilon in epsilons:
    print(f"epsilon={epsilon}")
    x_jacobi, iterations_jacobi = jacobi_method(a_vis, b, x0, epsilon)
    print(f"Метод Якобі: x={x_jacobi} кількість ітерацій={iterations_jacobi}")


epsilon=0.1
Метод Якобі: x=[-0.0237731887067178, 0.0298429364716831, 0.768381198524081, -0.0126813200276415, -0.0579920177105475, 0.216040171756044] кількість ітерацій=3
epsilon=0.01
Метод Якобі: x=[-0.0168710568169510, 0.0599513759685466, 0.808002092468245, 0.0287624805845212, 0.0348877383689195, 0.221172414622647] кількість ітерацій=7
epsilon=0.001
Метод Якобі: x=[-0.0343155047139909, 0.0756076130957632, 0.821688113543789, 0.0431099021886934, 0.0512753845245571, 0.222483445945924] кількість ітерацій=14
epsilon=1e-06
Метод Якобі: x=[-0.0369780494605123, 0.0792023777188896, 0.825153357011146, 0.0466723438645365, 0.0550185274698183, 0.223154799566071] кількість ітерацій=41


### EPSILON=0.1
#### МЕТОД ЯКОБІ: X=[-0.0237731887067178, 0.0298429364716831, 0.768381198524081, -0.0126813200276415, -0.0579920177105475, 0.216040171756044] 
#### КІЛЬКІСТЬ ІТЕРАЦІЙ=3

### EPSILON=0.01
#### МЕТОД ЯКОБІ: X=[-0.0168710568169510, 0.0599513759685466, 0.808002092468245, 0.0287624805845212, 0.0348877383689195, 0.221172414622647] 
#### КІЛЬКІСТЬ ІТЕРАЦІЙ=7

### EPSILON=0.001
#### МЕТОД ЯКОБІ: X=[-0.0343155047139909, 0.0756076130957632, 0.821688113543789, 0.0431099021886934, 0.0512753845245571, 0.222483445945924]
#### КІЛЬКІСТЬ ІТЕРАЦІЙ=14

### EPSILON=1e-06
#### МЕТОД ЯКОБІ: X=[-0.0369780494605123, 0.0792023777188896, 0.825153357011146, 0.0466723438645365, 0.0550185274698183, 0.223154799566071]
#### КІЛЬКІСТЬ ІТЕРАЦІЙ=41

In [156]:
def zeidel(A, b, x0, epsilon):
    n = A.shape[0]
    x = x0.copy()
    iterations = 0
    while True:
        iterations += 1
        for i in range(n):
            s = 0
            for j in range(n):
                if j != i:
                    s += A[i, j] * x[j]
            x[i] = (b[i] - s) / A[i, i]
        if max([abs(x[i]-x0[i]) for i in range(n)]) < epsilon:
            break
        x0 = x.copy()
    return x, iterations

### Розв'язки методом Зейделя

In [162]:
for epsilon in epsilons:
    print(f"epsilon={epsilon}")
    x_jacobi, iterations_jacobi = zeidel(a_vis, b, x0, epsilon)
    print(f"Метод Зейделя: x={x_jacobi} кількість ітерацій={iterations_jacobi}")


epsilon=0.1
Метод Зейделя: x=[0.0303310983233011, 0.0118655659192859, 0.762005090099602, -0.0203726435567576, -0.00281333511643204, 0.211782929411507] кількість ітерацій=2
epsilon=0.01
Метод Зейделя: x=[-0.0273659888757856, 0.0681611453331032, 0.815365517942873, 0.0370872796733325, 0.0457740584549829, 0.221477866232885] кількість ітерацій=5
epsilon=0.001
Метод Зейделя: x=[-0.0361569428392398, 0.0782590317337721, 0.824317630360592, 0.0458542637200683, 0.0542294712164703, 0.223011734049861] кількість ітерацій=9
epsilon=1e-06
Метод Зейделя: x=[-0.0369797763809341, 0.0792048099360541, 0.825155762401959, 0.0466748567838112, 0.0550213300392737, 0.223155299700437] кількість ітерацій=20


### EPSILON=0.1
#### МЕТОД ЗЕЙДЕЛЯ: X=[0.0303310983233011, 0.0118655659192859, 0.762005090099602, -0.0203726435567576, -0.00281333511643204, 0.211782929411507] 
#### КІЛЬКІСТЬ ІТЕРАЦІЙ=2

### EPSILON=0.01
#### МЕТОД ЗЕЙДЕЛЯ: X=[-0.0273659888757856, 0.0681611453331032, 0.815365517942873, 0.0370872796733325, 0.0457740584549829, 0.221477866232885]
#### КІЛЬКІСТЬ ІТЕРАЦІЙ=5

### EPSILON=0.001
#### МЕТОД ЗЕЙДЕЛЯ: X=[-0.0361569428392398, 0.0782590317337721, 0.824317630360592, 0.0458542637200683, 0.0542294712164703, 0.223011734049861] 
#### КІЛЬКІСТЬ ІТЕРАЦІЙ=9

### EPSILON=1e-06
#### МЕТОД ЗЕЙДЕЛЯ: X=[-0.0369797763809341, 0.0792048099360541, 0.825155762401959, 0.0466748567838112, 0.0550213300392737, 0.223155299700437] 
#### КІЛЬКІСТЬ ІТЕРАЦІЙ=20

# Завдання 5 
## Скласти програму для реалізації методу верхньої релаксації. Провести серію чисельних експериментів із варіацією значень εε, початкових наближень і релаксаційного параметру ωω. За допомогою спроб визначити оптимальне значення параметру ωω.

In [166]:
a_vis.is_symmetric()

False

### Бачимо, що наша матриця не симетрична, тож маємо провести її симетризацію задля забезпечення збіжності.

In [171]:
A = (a_vis + a_vis.T)/2
A.is_symmetric()

True

In [180]:
def upper_relaxation_method(A, b, w, x0, epsilon, max_iter):
    n = len(x0)
    x = x0.copy()
    num_iters = 0
    for _ in range(max_iter):
        num_iters += 1
        x_new = x.copy()
        for i in range(n):
            s = sum(-A[i, j] * x_new[j] for j in range(n) if i != j)
            x_new[i] = (1 - w) * x[i] + w * (b[i] + s) / A[i, i]
        if max(abs(x_new[i] - x[i]) for i in range(n)) < epsilon:
            return x_new, num_iters
        x = x_new
    return x, num_iters

### Ров'язок методом верхньої релаксації для w = 0.1, epsilon = 0.1, max_iter = 100(як приклад)

In [183]:
upper_relaxation_X = upper_relaxation_method(A, b, 0.1, x0, 0.1, 100)
Matrix(upper_relaxation_X[0])

Matrix([
[  0.00229885057471264],
[ -0.00660099950024988],
[   0.0787550382414427],
[-0.000145954572723296],
[  0.00253849492321749],
[  0.00508197638463093]])

### Тепер визначимо оптимальні значення параметрів epsilon та omega.

In [185]:
omegas = [0.1, 0.5, 0.9, 1.1, 1.5, 1.9] 
max_iter = 1000

In [187]:
optimal_omega = None
optimal_epsilon = None
min_iterations = float("inf")


for epsilon in epsilons:
    print(f"Точність: {epsilon}")
    for omega in omegas:
        x, iters = upper_relaxation_method(A, b, omega, x0, epsilon, max_iter)
        print(f"omega = {omega}, розв'язок: {x}, кількість ітерацій: {iters}")
        
        if iters < min_iterations:
            min_iterations = iters
            optimal_omega = omega
            optimal_epsilon = epsilon


print(f"Оптимальний варіант: omega = {optimal_omega}, точність = {optimal_epsilon}, кількість ітерацій: {min_iterations}")


Точність: 0.1
omega = 0.1, розв'язок: [0.00229885057471264, -0.00660099950024988, 0.0787550382414427, -0.000145954572723296, 0.00253849492321749, 0.00508197638463093], кількість ітерацій: 1
omega = 0.5, розв'язок: [-6.02699272803781e+144, 2.59774712984977e+144, 2.21284561361979e+144, 2.32656284359586e+144, 2.89005904833345e+144, 1.72432482275975e+143], кількість ітерацій: 1000
omega = 0.9, розв'язок: [-4.21026462142550e+314, 2.07948399666377e+314, 1.87226303961340e+314, 1.97930688888182e+314, 2.54780652562965e+314, 1.63320645817260e+313], кількість ітерацій: 1000
omega = 1.1, розв'язок: [-8.18034062937528e+413, 4.38857914145139e+413, 4.07047759104647e+413, 4.31475109804635e+413, 5.68463638911010e+413, 3.78848649367801e+412], кількість ітерацій: 1000
omega = 1.5, розв'язок: [-6.21705285403821e+620, 3.98824745908361e+620, 3.91319538765784e+620, 4.16678511243201e+620, 5.79810016608688e+620, 4.14484137549822e+619], кількість ітерацій: 1000
omega = 1.9, розв'язок: [-2.36556331011289e+816, 1

### Оптимальний варіант: omega = 0.1, точність = 0.1, кількість ітерацій: 1